### PR TITLE
Support hyperlinked titles according to draft of CSL v1.0.2 spec

### PR DESCRIPTION
**Pull request by benrbray**
As described in #86, the working draft of the CSL v1.0.2 specification has a [recommendation for automatically linking titles](https://github.com/citation-style-language/documentation/blob/master/specification.rst#appendix-vi-links) when the citation style does not explicitly render the URL.  See also the [discussion by the CSL maintainers](https://github.com/citation-style-language/documentation/issues/76).  This draft PR adds this functionality to `citeproc` (and transitively to `pandoc`).

### Relevant Snippet from the draft CSL v1.0.2 Spec

> The CSL syntax does not have support for configuration of links. However, processors should include links on bibliographic references, using the following rules:
>
> If the bibliography entry for an item renders any of the following identifiers, the identifier should be anchored as a link, with the target of the link as follows:
>
> 1. ``url``: output as is
> 2. ``doi``: prepend with "https://doi.org/"
> 3. ``pmid``: prepend with "https://www.ncbi.nlm.nih.gov/pubmed/"
> 4. ``pmcid``: prepend with "https://www.ncbi.nlm.nih.gov/pmc/articles/"
>
> If the identifier is rendered as a URI, include rendered URI components (e.g. "https://doi.org/") in the link anchor. Do not include any other affix text in the link anchor (e.g. "Available from: ", "doi: ", "PMID: ").
> If the bibliography entry for an item does not render any of the above identifiers, then set the anchor of the link as the item ``title``. If ``title`` is not rendered, then set the anchor of the link as the full bibliography entry for the item. Set the target of the link as one of the following, in order of priority:
> 
> 1. ``doi``: prepend with "https://doi.org/"
> 2. ``pmcid``: prepend with "https://www.ncbi.nlm.nih.gov/pmc/articles/"
> 3. ``pmid``: prepend with "https://www.ncbi.nlm.nih.gov/pubmed/"
> 4. ``url``: output as is
> 
> If the item data does not include any of the above identifiers, do not include a link.
> 
> Citation processors should include an option flag for calling applications to disable bibliography linking behavior.

### Tasks / Considerations

Finishing the Feature

- [x] Currently my PR only supports hyperlinking *titles*, but the CSL spec says that the *entire* bibitem should be hyperlinked when neither a title nor explicit identifier element is present.  I see a few ways to accomplish this, **what do you think is best?**  
    * Initially, I wanted to wrap the output of `Eval.hs/evalItem` with a tag, but the `Tagged` constructor only takes a single `Output a`, instead of a list `[Output a]`.
    * (Option 1) So, one option would be to add a new `Link Identifier [Output a]` constructor to the `Output` type.
    * (Option 2) Alternatively, instead of using tags, another field called `formatLink :: Maybe Text` could be added to the `Formatting` type.  When this field is not `Nothing`, the `addFormatting` would render the link using `addHyperlink`.  However, this would require changing the `Element.hs/getFormatting` function to handle links, and I'm not sure if that makes sense.
    * **Decision:**  Chose Option 1
- [x] Currently, URLs etc are linkified in `pandoc`.  It should be straightforward to move this functionality into the `Eval.hs/eText` function.  **Should I do it?**
   * **Decision:**  Yes, move the linkify code from pandoc to citeproc
- [x] If only an ISBN is provided (but none of the other `Identifier` types) , should it be used to hyperlink the title?  This isn't in the CSL spec, but since `pandoc` already linkifies ISBNs, it would be natural to include here.
    * **Decision:**  Don't linkify ISBN anymore

Testing / Documentation

- [x] I compiled `pandoc` against my local copy of `citeproc` and everything seems to work as expected, but **further testing is needed**.
- [x] What would be a good way to **write tests** for this feature?  Currently the `citeproc` tests all use the `CslJson` data type, which [does not support hyperlinks](https://github.com/jgm/citeproc/blob/master/src/Citeproc/CslJson.hs#L150).  I suppose we could add another `CiteprocOutput` instance just for testing.
- [x] What should the **default** be for the new `linkTitles` option?
    * `False` would be more backward-compatible, but I'd argue `True` is more in line with what users might expect.  This has been a pet peeve of mine for a while (when viewing HTML output by pandoc) and it's very satisfying to finally be able to fix it :).
    * **Decision:** `linkTitles` should default to False
- [x] Update `citeproc` documentation to reflect this behavior, in particular the new `linkTitles` option.
    * What should the default for `linkTitles` be?  I think setting it to `True` in the context of the Pandoc HTML Writer would be appropriate, but I
- [x] Consider how these changes will impact `pandoc`.  Probably a change to `pandoc` will be required, at the very least to document this functionality and set a sensible default for the `linkTitles` option.

Thanks for your feedback!

### CODE REVIEW

File: src/Citeproc/Eval.hs [hunk 1]
```diff
@@ -1142,16 +1148,41 @@ evalItem layout (position, item) = do
...
+                 linkTitle _ x = x
+             
+             -- when no links were rendered, hyperlink title, if it exists
+             usedLink <- gets stateUsedIdentifier
+
+             let xs' = case mbident of
+                         Nothing     -> xs
+                         Just ident  -> if usedLink
+                                           then xs
+                                           else fmap (transform (linkTitle ident)) xs
```
Comment (benrbray) [inline src/Citeproc/Eval.hs:1176]: Currently `transform` is called twice -- one for `linkTitle` and one for `addLangtoFormatting`.  Should these two uses be merged into one?

File: src/Citeproc/Types.hs [hunk 2]
```diff
@@ -141,17 +142,23 @@ import Citeproc.Unicode (Lang(..), parseLang, renderLang)
...
+    -- add a hyperlink to the title instead.
   }
   deriving (Show, Eq)
 
 defaultCiteprocOptions :: CiteprocOptions
 defaultCiteprocOptions =
   CiteprocOptions
-  { linkCitations = False }
+  { linkCitations = False 
+  , linkTitles    = True
```
Comment (benrbray) [inline src/Citeproc/Types.hs:160]: Should the default for `linkTitles` be `True` or `False`?

Comment (jgm) [reply src/Citeproc/Types.hs:160]: Maybe it should be False for now.  If we decide to implement `addHyperlink` for CslJson, then having this as True would affect automated tests.  Pandoc can always set this to True if it wants this by default.

File: src/Citeproc/Eval.hs (hunk 1 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:1176]: If it can be done without contorting the code, this seems sensible.

**Comment by jgm** [comment]
As to some of your questions:

- Yes, I think we should move linkifying of URL elements from pandoc to citeproc.
- As for ISBNs, if they're not part of the CSL suggestion, let's stop linkifying them.
- On adding links to the whole citation, I'm leaning towards the option of adding a Link constructor to `data Output`.  (That way Formatting keeps its association with the formatting that can be specified via attributes in a CSL stylesheet...)
-  Testing:  not sure.  We could always just implement `addHyperlink` for CslJson.  It would be a bit nonstandard, but presumably that's the direction CslJson will be headed when the revised spec comes out? I guess the problem is that this would make existing tests fail.  But we could prevent that if we added an explicit option for linkifying URLs.  With this option turned off, the `addHyperlink` should never be activated.

**Comment by benrbray** [comment]
Thanks for the feedback, I pushed another commit that checks off a few boxes.  Some notes / things that came up, in addition to the code comments:

- [ ] Some styles (for example, the [Chicago style]) specify that DOIs (for instance) should have a prefix that turns the DOI in the URL.  Should we handle this somehow or let users work it out themselves with some combination of config options?
- [ ] I started moving the linkify code out of pandoc, trying to preserve the behavior as closely as possible.  However I think there are some differences, as Pandoc operated on the variables themselves by replacing URLs with a `FancyVal Inlines`.  In `citeproc` we need to go through `addHyperlink` instead, which happens at the very end of citation processing.
- [ ] The new changes cause some `citeproc` tests to fail.  However, quite a few tests are expected to fail under normal circumstances, so it's hard to identify from the log output which subset of tests were expected to succeed, but didn't.  Any suggestions?

File: src/Citeproc/Eval.hs [hunk 3]
```diff
@@ -1162,18 +1162,30 @@ evalItem layout (position, item) = do
...
-                         Just ident  -> if usedLink
-                                           then xs
-                                           else fmap (transform (linkTitle ident)) xs
+             usedLink  <- gets stateUsedIdentifier
+             usedTitle <- gets stateUsedTitle
+             inBiblio  <- asks contextInBibliography 
+
+             -- when no links were rendered for a bibliography item, hyperlink
+             -- the title, if it exists, otherwise hyperlink the whole item
+             let xs' =
```
Comment (benrbray) [inline src/Citeproc/Eval.hs:1177]: I'll try to clean up this nested if/else later

File: src/Citeproc/Types.hs [hunk 4]
```diff
@@ -146,18 +148,26 @@ data CiteprocOptions =
   CiteprocOptions
```
Comment (benrbray) [inline src/Citeproc/Types.hs:148]: thinking of adding a `linkPrefixes` option to CiteprocOptions to control whether or not URL prefixes are added to links (e.g. `https://doi.org/` in front of a DOI)

File: src/Citeproc/Types.hs [hunk 5]
```diff
@@ -1438,6 +1448,7 @@ removeEmptyStrings = filter (not . isEmptyString)
 
 data Output a =
     Formatted Formatting [Output a]
+  | Linked LinkType Text [Output a]
```
Comment (benrbray) [inline src/Citeproc/Types.hs:1451]: Keeping track of the `LinkType` as part of the output was necessary so that the appearance of links can be controlled by flags in `CiteprocOptions`.

File: src/Citeproc/Types.hs [hunk 6]
```diff
@@ -1454,28 +1466,42 @@ instance Uniplate (Output a) where
...
 data Identifier =
       IdentDOI Text
     | IdentPMCID Text
     | IdentPMID Text
     | IdentURL Text
   deriving (Show, Eq)
 
-identifierToUrl :: Identifier -> Text
-identifierToUrl ident =
+identifierToURL :: Identifier -> Text
```
Comment (benrbray) [inline src/Citeproc/Types.hs:1484]: identifierToURL is adapted from pandoc's linkify function, and I tried to preserve the same behavior

File: src/Citeproc/Eval.hs [hunk 7]
```diff
@@ -2293,6 +2319,8 @@ getDisplayName nameFormat formatting order name = do
                             case formatSuffix f of
                               Just t | endsWithSpace t -> Nothing
                               _ -> Just " " } [formatted f x, y]
+      Linked lt i x <+> y =
+        formatted mempty{ formatDelimiter = Just " " } [Linked lt i x, y]
```
Comment (benrbray) [inline src/Citeproc/Eval.hs:2323]: does this implementation of `<+>` make sense?  It wasn't clear to me what the intended meaning is

**Comment by jgm** [comment]
On the `linkTitles` option:  see #85. I'm now thinking the simplest thing to do would be to set it in pandoc to true if `linkCitations` is true.  This would make hyperlinks from citations to the bibliography and hyperlinks from the bibliography to external resources go together.  Is there reason to make these independent?

**Comment by jgm** [comment]
> Some styles (for example, the [Chicago style]) specify that DOIs (for instance) should have a prefix that turns the DOI in the URL. Should we handle this somehow or let users work it out themselves with some combination of config options?

The current citeproc code does insert these prefixes.  We need to make sure that when we add the links, the link anchor includes these prefixes, otherwise it looks funny. This currently works properly.

> I started moving the linkify code out of pandoc, trying to preserve the behavior as closely as possible. However I think there are some differences, as Pandoc operated on the variables themselves by replacing URLs with a FancyVal Inlines. In citeproc we need to go through addHyperlink instead, which happens at the very end of citation processing.

Now that you have a Link type in Output, you should just be able to use that when resolving one of these URL variables, no?

> The new changes cause some citeproc tests to fail. However, quite a few tests are expected to fail under normal circumstances, so it's hard to identify from the log output which subset of tests were expected to succeed, but didn't. Any suggestions?

I use `make test` to run the tests; this puts the log in spec.log.  So, you can run with master, then `cp spec.log spec.master.log`, then run with your branch, then `diff spec.master.log spec.log`.  It's still a bit confusing because you have a diff of a log that itself contains diffs, but that's what I do...

**Comment by jgm** [comment]
Suggestion:  now that you have the Linked Output constructor, you can revert the addition of TagTitle.  Instead of tagging the title as TagTitle, you can just slap a Linked on it at the point where the title is resolved. Similarly with the URL variables like `doi`.  (Of course, you'll need to check the config option to make sure these links are called for.)  Then, at the end, you can simply traverse to see if the whole entry contains any Linked constructors, and add one to the whole thing if not.

That seems cleaner to me.  (And it also gets the behavior right -- remember that the whole entry is to be linked only if there is no linked title *or* URL fields.)

File: src/Citeproc/Eval.hs (hunk 7 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:2323]: Yeah, that's the sort of thing one wishes the author had documented in the code.
(Even the author wishes this.)
If memory serves, these are used to format names.  `<:>` goes where a comma (or whatever the separator is) will go and `<+>` where a space will go.
IT shouldn't matter much for Linked, since this won't be used in names, but we should still try to get it right.
I think you have.

File: src/Citeproc/Types.hs (hunk 4 above)
Comment (jgm) [reply src/Citeproc/Types.hs:148]: Isn't this something that is handled in the style itself?

File: src/Citeproc/Types.hs (hunk 5 above)
Comment (jgm) [reply src/Citeproc/Types.hs:1451]: Why do we need to distinguish these?
I would have thought that any Linked would simply be rendered as a hyperlink to the given URL (Text) with the `[Output a]` as anchor.

Comment (jgm) [reply src/Citeproc/Types.hs:1451]: Oh, I see -- the point is not to affect *how* it is rendered, but whether the link is rendered at all.
Couldn't this be done earlier, though?  Don't insert the Linked constructor in the first place unless the options call for that sort of link.

File: src/Citeproc/Types.hs (hunk 4 above)
Comment (benrbray) [reply src/Citeproc/Types.hs:148]: Oh, I get it now, thanks.  You're saying that the CSL itself decides whether the DOI should be prefixed with `https://doi.org/` or `DOI:` or something else.  We should only be changing the link target, not the anchor.  This is probably the reason for some of the failing tests in my most recent commit.

File: src/Citeproc/Types.hs (hunk 5 above)
Comment (benrbray) [reply src/Citeproc/Types.hs:1451]: Yeah, technically the decision whether to create a link or not could be moved to [`Eval.hs/evalItem`](https://github.com/jgm/citeproc/blob/master/src/Citeproc/Eval.hs#L1128) where the `Linked` constructor is actually used.  However, we'd need to pass the `CiteprocOptions` along into the evaluation [`Context`](https://github.com/jgm/citeproc/blob/master/src/Citeproc/Eval.hs#L41), since they aren't currently available inside the eval functions.  Is that fine with you?

Comment (jgm) [reply src/Citeproc/Types.hs:1451]: Ah, I see.  Eval has no access to CiteprocOptions.  Well, I think the way you're doing it is okay, then.

**Comment by jgm** [comment]
> Suggestion: now that you have the Linked Output constructor, you can revert the addition of TagTitle. Instead of tagging the title as TagTitle, you can just slap a Linked on it at the point where the title is resolved.

I now want to go back on this. Using TagTitle makes sense here, since we can't decide whether to linkify the title until we've rendered the whole citation and determined whether it contains any Linked items.  (So, the way you have it seems fine to me.)

File: src/Citeproc/Eval.hs [hunk 8]
```diff
@@ -1542,10 +1547,10 @@ eText (TextVariable varForm v) = do
                       coerce (referenceId ref)
             return NullOutput
 
-    "DOI"   -> handleIdent IdentDOI
-    "PMCID" -> handleIdent IdentPMCID
-    "PMID"  -> handleIdent IdentPMID
-    "URL"   -> handleIdent IdentURL
+    "DOI"   -> handleIdent fixShortDOI IdentDOI
```
Comment (benrbray) [inline src/Citeproc/Eval.hs:1550]: Passing `fixShortDOI` here so that the link anchor for a short DOI like `10/abcde` appears as `abcde` instead.

File: src/Citeproc/Eval.hs [hunk 9]
```diff
@@ -1186,15 +1190,17 @@ evalItem layout (position, item) = do
                                         then fmap (transform (linkTitle url)) xs
                                         -- hyperlink the entire bib item
                                         else [Linked LinkBibItem url xs] 
+            
+             let xs'' = fmap (transform fixLinkPrefixes) xs'
```
Comment (benrbray) [inline src/Citeproc/Eval.hs:1194]: This line is needed because an element like `<text variable="DOI" prefix="https://doi.org/">` will be realized as a `Formatted (Formatting ..) [Linked ..]` during citation processing.  Since the `prefix` is part of the formatting, this means that without this line, you'd get an ugly link like `https://doi.org/`[`12345`](https://doi.org/12345) where only the DOI itself is part of the anchor.

It makes me think that maybe we should've added links to the `Formatting` type instead, so we can control exactly when they're rendered.  That would avoid the need to traverse the tree again.

One thing to note is that with the fix applied, elements like `<text variable="DOI" prefix="DOI: ">`, will render as **[DOI: 12345](https://doi.org/12345)** instead of **DOI: [12345](https://doi.org/12345)**.

**Comment by benrbray** [comment]
Locally, tests are no longer failing.  I think my misunderstanding of how prefixes should be handled was to blame.  Fixed now!

File: src/Citeproc/Eval.hs (hunk 3 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs:1181]: I removed one level of nesting, which I think is good enough unless you have other suggestions.  Other attempts to clean it up with `<|>` / `msum` etc I think made it harder to read and required defining some functions that wouldn't be used anywhere else.

File: src/Citeproc/Eval.hs (hunk 9 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:1192]: I think the way pandoc currently works is that these prefixes are only incorporated into the link if do not end with a space. (See `fixLinks` in T.P.Citeproc and related issue jgm/pandoc#6723).  That would be the best behavior.  Can we do that?

If we put the link in Formatting, then automatically *every* prefix *and* suffix would be linked, which is not what we want, I think.

File: src/Citeproc/Eval.hs (hunk 8 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:1548]: Is this the desired behavior?  Where is this documented?  Does pandoc currently do this or is it a behavior change?

Comment (benrbray) [reply src/Citeproc/Eval.hs:1548]: I believe this is how pandoc currently works.  I had to squint at [`linkifyVariables`](https://github.com/jgm/pandoc/blob/82638ad53bb47acde835083390f8211ac21cda6c/src/Text/Pandoc/Citeproc.hs#L541-L547) for a while earlier to realize it (note `fixShortDOI x` is passed to the `tolink` function, which determines the link anchor/target).  I also just confirmed this behavior with the release copy of pandoc. 

As for whether this is desirable:  I'd say yes, since every shortDOI `10/abcde` corresponds to the real DOI `abcde`, right? 
 If we don't do it this way, and we have a `<text variable="doi" prefix="https://doi.org/".`, the link will render incorrectly as `https://doi.org/10/abcde` when given a shortDOI `10/abcde`.  Instead, it should be `https://doi.org/abcde`.

File: src/Citeproc/Eval.hs (hunk 9 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs:1192]: Hmm, interesting.  

It would be easy to wrap the `Formatting` in a `Link` only when the prefix has a trailing space.  Note that this would also automatically wrap any suffix as well (as does the current code).

The [`fixLinks`](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Citeproc.hs#L424) function actually does something more clever, though, right?  It only acts when the prefix + anchor equals the link target.  That seems a bit more difficult to reproduce here.

**Comment by benrbray** [comment]
> On the `linkTitles` option: see #85. I'm now thinking the simplest thing to do would be to set it in pandoc to true if `linkCitations` is true. This would make hyperlinks from citations to the bibliography and hyperlinks from the bibliography to external resources go together. Is there reason to make these independent?

I don't think it's inconceivable that someone could want to link titles but not citations, or citations but not titles.  It's difficult for me to say, since I don't have a good sense of the scope of how people use pandoc in the real world.  For my current use case, as part of the markdown -> HTML conversion process in Hakyll, I would want to use the configuration `linkCitations=True, linkTitles=True, linkBibItems=False`.

File: src/Citeproc/Eval.hs (hunk 8 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:1548]: OK, right.

File: src/Citeproc/Eval.hs (hunk 9 above)
Comment (jgm) [reply src/Citeproc/Eval.hs:1192]: Oh right, I forgot that `fixLinks` does that.  I think we should reproduce that behavior.
Your function seems good to me.

**Comment by jgm** [comment]
`linkBibItems=False` -- just curious why?  This would only kick in when the bibliography item has no title, and contains a URL, but this is not rendered.  I imagine such cases are fairly rare (most entries have titles), but wouldn't you want the link in such cases?

**Comment by benrbray** [comment]
> wouldn't you want the link in such cases?

Purely aesthetic reasons -- it's kind of ugly when the link surrounds the whole bibliography entry.  But in that case I would probably just add a proper title to the bibliography item.  So I agree it's probably a pathological case, and we wouldn't lose very much by combining all three.  Either way is fine with me.

File: src/Citeproc/Eval.hs (hunk 9 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs:1199]: Ok!  Just pushed the fix.

File: src/Citeproc/Eval.hs [hunk 10]
```diff
@@ -1142,16 +1148,64 @@ evalItem layout (position, item) = do
...
+                      ]
+             let mburl = identifierToURL <$> mbident
+            
+             -- hyperlink any titles in the output
+             let linkTitle url (Tagged TagTitle x) = Linked LinkTitle url [x]
+                 linkTitle _ x = x
+             
+            -- ensure correct handling of link prefixes like (https://doi.org/)
+            -- when a link's prefix+anchor=target, ensure the link includes the prefix
+            -- (see pandoc#6723 and citeproc#88; also pandoc's fixLinks function)
```
Comment (jgm) [inline src/Citeproc/Eval.hs:1173]: Presumably we'll remove fixLinks from pandoc after this is added here, so referring to it in the comment may be confusing.

File: src/Citeproc/Eval.hs [hunk 11]
```diff
@@ -1142,16 +1148,64 @@ evalItem layout (position, item) = do
...
+                      [ IdentDOI   <$> (valToText =<< lookupVariable "DOI" ref)
+                      , IdentPMCID <$> (valToText =<< lookupVariable "PMCID" ref)
+                      , IdentPMID  <$> (valToText =<< lookupVariable "PMID" ref)
+                      , IdentURL   <$> (valToText =<< lookupVariable "URL" ref)
+                      ]
+             let mburl = identifierToURL <$> mbident
+            
+             -- hyperlink any titles in the output
+             let linkTitle url (Tagged TagTitle x) = Linked LinkTitle url [x]
+                 linkTitle _ x = x
```
Comment (jgm) [inline src/Citeproc/Eval.hs:1169]: Here's a thought:  instead of `Linked LinkTitle url [x]`, what about `Linked url [Tagged TagTitle x)`?  The extra parameter on `Linked` wouldn't really be needed, since we could just check to see if the content is a title.

File: src/Citeproc/Eval.hs [hunk 12]
```diff
@@ -1142,16 +1148,64 @@ evalItem layout (position, item) = do
...
+             
+            -- ensure correct handling of link prefixes like (https://doi.org/)
+            -- when a link's prefix+anchor=target, ensure the link includes the prefix
+            -- (see pandoc#6723 and citeproc#88; also pandoc's fixLinks function)
+             let fixLinkPrefixes (Formatted f [Linked lt u ys]) 
+                                 | checkPrefix f u ys = Linked lt u [Formatted f ys]
+                 fixLinkPrefixes y = y
+                 checkPrefix f url ys =
+                    let anchor = mconcat mempty (map outputToText ys)
```
Comment (jgm) [inline src/Citeproc/Eval.hs:1178]: I'm confused by the `mempty` here.  Why isn't it `mconcat (map outputToText ys)`?

File: src/Citeproc/Eval.hs [hunk 13]
```diff
@@ -73,6 +73,10 @@ data EvalState a =
   , stateRefMap         :: ReferenceMap a
   , stateReference      :: Reference a
   , stateUsedYearSuffix :: Bool
+  , stateUsedIdentifier :: Bool
+  -- ^ tracks whether an identifier (DOI,PMCID,PMID,URL) has yet been used
+  , stateUsedTitle      :: Bool
+  -- ^ tracks whether the item title has yet been used
```
Comment (jgm) [inline src/Citeproc/Eval.hs:79]: Do we really need these?
We could traverse to see if we have a Linked or a TagTitle.
If we have a Linked, then it means we had one of these identifiers and we're done.
If no Linked, but TagTitle, then we can put a Linked on the title.
If none of the above, put a Linked on the whole thing.

Comment (jgm) [reply src/Citeproc/Eval.hs:79]: I suppose it's slightly more efficient to track this in state, so maybe that justifies this.

File: src/Citeproc/Eval.hs [hunk 14]
```diff
@@ -1142,16 +1148,64 @@ evalItem layout (position, item) = do
...
+                     then xs
+                   else case mburl of
+                         Nothing  -> xs
+                         Just url -> if usedTitle
+                                        -- hyperlink the title
+                                        then fmap (transform (linkTitle url)) xs
+                                        -- hyperlink the entire bib item
+                                        else [Linked LinkBibItem url xs] 
+            
+             let xs'' = fmap (transform fixLinkPrefixes) xs'
```
Comment (jgm) [inline src/Citeproc/Eval.hs:1199]: Presumably we only need this transformation when `usedLink && inBiblio`?  So we could localize it more and avoid unnecessary work?

Or, maybe instead of having a traversal with `fixLinkPrefixes`, we could just build this logic into `renderOutput` (in Types)?

File: src/Citeproc/Eval.hs (hunk 12 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs]: Good catch, I must've been thinking about a `fold` and the type inference was a bit too generous.  Fixed.

File: src/Citeproc/Eval.hs (hunk 10 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs]: Sure, I removed it.

**Comment by benrbray** [comment]
While testing, I noticed that `citeproc` currently overrides `linkCitations` with `False` when processing the bibliography.  Is this important, or can we remove the override?

If we can't remove the override, then we'll need to keep `linkCitations` separate from whatever option(s) we include for url/title/bibitem hyperlinking.

https://github.com/jgm/citeproc/blob/0c267f3e7580ed22f98792ad6f2ceb7b47402e77/src/Citeproc.hs#L42-L47

File: src/Citeproc/Eval.hs (hunk 14 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs]: Seems reasonable, so I made the change.

File: src/Citeproc/Eval.hs (hunk 11 above)
Comment (benrbray) [reply src/Citeproc/Eval.hs:1169]: Seems reasonable, so I made the change.

**Comment by benrbray** [comment]
Ok, my latest commits include test cases for the feature as described in the draft spec as well as (hopefully) most of the edge cases we discussed.  I also made it possible for test cases to specify `CiteprocOptions` in a JSON format.  I also took your suggestions about simplifying the code and combining `linkURLs`, `linkTitles`, and `linkBibItems` all into one `linkBibliography` option.

**Comment by jgm** [comment]
This all looks great! I don't think we need a CslJson parser for links.  The links are going to be added in bibliography processing, not passed in the bib entries themselves.

**Comment by jgm** [comment]
It's saying there's a merge conflict with src/Citeproc/CslJson.hs that still needs to be resolved. Can you have a look at that?

**Comment by benrbray** [comment]
Merge conflict resolved!  I think we have the following todos, let me know if there's anything else:

- [ ] decide if `linkCitations` should be merged with `linkBibliography` and decide on default values
- [ ] update citeproc documentation
- [ ] apply the necessary changes to pandoc and update pandoc documentation accordingly
    * update the options [passed to citeproc](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Citeproc.hs#L76) in `pandoc/Text.Pandoc.Citeproc.hs`
    * remove the [`linkifyVariables`](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Citeproc.hs#L526-L547) function in `pandoc/Text.Pandoc.Citeproc.hs`
    * remove [`fixLinks`](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Citeproc.hs#L423-L428) in `pandoc/Text.Pandoc.Citeproc.hs`
    * add `CslLink` to the [`fromCslJson` instance](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Citeproc/CslJson.hs#L14) in `pandoc/Text.Pandoc.Citeproc/CslJson.hs`

I have a local copy of `pandoc` with these changes and I can start a PR over there.  Right now I'm using `cabal.project.local` to compile against the changes to `citeproc`.

**Comment by jgm** [comment]
I'm having a hard time thinking of a good reason someone might have to want to separate linking citations to the bibliography and linking the bibliography to external resources.  So I'm inclined to just make both of these controlled by `linkCitations` in pandoc -- of course, we could always keep the options separate in `citeproc`, which opens up the possibility of doing things differently in the future.

What do you think?

**Comment by jgm** [comment]
> decide if linkCitations should be merged with linkBibliography and decide on default values

I think keeping them separate in citeproc makes sense, and defaulting to False makes sense; that's what you have here. We can decide separately what values pandoc should set them to by default, and whether pandoc should couple them.

>  update citeproc documentation

Do you mean the README.md or man/citeproc.1.md?  Does your patch change the `citeproc` executable at all?  (Perhaps we could add a command-line option to set `link-citations` and `link-bibliography` -- but currently we don't have one for `link-citations`, so I think this could be considered a separate project.)

The README could use a few more words in the description of CiteprocOptions, but I think only a few are needed.

> apply the necessary changes to pandoc 

I saw the PR about this, thanks.  Let's get the citeproc changes merged, then handle those.  Is this ready to go as far as you're concerned?

**Review by jgm — APPROVED** [review]

**Comment by benrbray** [comment]
Yeah I think this is ready to go!  

* I am also inclined to keep `linkCitations` and `linkBibliography` separate, for now.  
* As long as `linkBibliography` defaults to `False`, I think the behavior of the executable is the same, so yeah, no change to those docs is needed for now
* I added a description of `linkBibliography` to the readme.

Thanks for your helpful feedback on this PR!

File: README.md [hunk 15]
```diff
@@ -40,8 +40,17 @@ integrated citation support and bibliography format conversion
...
-  to set whether citations are hyperlinked to the bibliography)
+- a `CiteprocOptions` structure, which includes the following options:
+
+  * `linkCitations` controls whether citations are hyperlinked 
+     to the bibliography.
+
+  * `linkBibliography` automatically linkifies any identifiers (DOI, 
+     PMCID, PMID, or URL) appearing in a bibliography entry.  When
+     an entry has a DOI, PMCID, PMID, or URL available (in order of 
+     priority) but the style does not explicitly render at least one
```
Comment (jgm) [inline README.md:51]: I wasn't sure what was meant here by "in order of priority."  Why does the order matter?  We're just checking to see if [one of these is present, but none are rendered], right?

Comment (jgm) [reply README.md:51]: Should we mention or link to the CSL recommendation here?

Comment (benrbray) [reply README.md:51]: By "in order of priority", I mean that if an entry specifies more than one identifier (DOI, PMCID, PMID, URL), the target of the hyperlink will be chosen in that order of priority.  Is there some way I can make the wording more clear?  And sure, linking the CSL recommendation makes sense.

Comment (jgm) [reply README.md:51]: I get it now.  Maybe it's okay as it stands, but perhaps clearer would be:

When an entry has a DOI, PMCID, PMID, or URL available but none of these are rendered by the style, add a link to the title (or, if no title is present, the whole entry), using the URL for the DOI, PMCID, PMID, or URL (in that order of priority).

Comment (benrbray) [reply README.md]: That is more clear, thanks!  I committed it along with a link to the spec.

**Comment by jgm** [comment]
Whoops, I meant to squash those commits into one and pressed the wrong button. Oh, well.

**Comment by jgm** [comment]
By the way, I was puzzled by that `mconcat mempty [..list...]` which type checked.
```
λ> mconcat mempty ["a","b","c"]
()
```
What is the type of `mempty` here???

**Comment by benrbray** [comment]
It was something like this!  I guess the type checker wanted to make sure I was paying attention :)

```
x :: String
x = mconcat mempty ["a", "b", "c"]
-- mconcat :: [[[Char]] -> String] -> [[Char]] -> String
-- mempty :: [[[Char]] -> String]
```

**Comment by jgm** [comment]
I didn't realize Monoid instances were defined for functions like that!  Neat.
